File: cpp/Array/Array.cpp
#include <iostream>
#include "Array.h"

using namespace std;


bool Array::isFull() {
    return count == sizeof(items) / sizeof(items[0]);
}

bool Array::isEmpty() {
    return count == 0;
}

void Array::resize() {
    int *newItems = new int[count * 2];
    for(int i=0; i<count; i++)
        newItems[i] = items[i];

    delete [] items;
    items = newItems;
}

void Array::insert(int item) {
    if(isFull()) resize();
    items[count++] = item;
}

void Array::insertAt(int item, int index) {
    if(isFull()) resize();
    for(int i = count-1; i >= index; i--)
        items[i + 1] = items[i];
    items[index] = item;
    count++;
}

void Array::remove() {
    items[count] = 0;
    count--;
}

void Array::removeAt(int index) {
    for(int i = index; i < count; i++)
        items[i] = items[i + 1];
    count--;
}

void Array::reverse() {
    int* newItems = new int[count];
    for(int i = 0; i < count; i++)
        newItems[i] = items[count - i - 1];
    items = newItems;
}

void Array::print() {
    for(int i = 0; i < count; i++)
        cout << items[i] << " ";
    cout << endl;
}

int Array::indexOf(int item) {
    for(int i = 0; i < count; i++)
        if (items[i] == item)
            return i;
    return -1;
}


Array Array::intersect(Array other) {
    Array intersection;
    for(int i = 0; i < count; i++) {
        if (other.indexOf(items[i]) >= 0)
            intersection.insert(items[i]);
    }
    return intersection;
}

File: python/Array/Array.py

class DynamicArray:
    def __init__(self):
        self.count = 0 
        self.capacity = 1 
        self.items = self.make_array(self.capacity) 
        
    
    def __resize(self):
        newItems = self.make_array(2 * self.capacity) 
          
        for i in range(self.count):  
            newItems[i] = self.items[i] 
              
        self.items = newItems  
        self.capacity = 2 * self.capacity 
        
        
    def __isFull(self):
        return self.count == self.capacity
    
    
    def __isEmpty(self):
        return self.count == 0
    
    
    def __size(self):
        return self.count
         
         
    def insert(self, item):
        if(self.__isFull()): self.__resize()
            
        self.items[self.count] = item  
        self.count += 1
    
    
    def insertAt(self, item, index):
        if index<0 or index>self.count:
            return -1
        if self.__isFull(): self.__resize()
        for i in range(self.count-1, index-1, -1):
            self.items[i+1] = self.items[i]
        self.items[index] = item
        self.count += 1
       
        
    def reverse(self):
        newItems = self.make_array(2 * self.capacity) 
        for i in range(self.count):
            newItems[i] = self.items[self.count - i - 1]
        
        self.items = newItems
    
    
    def max(self):
        max_num = 0
        for i in range(self.count):
            if self.items[i] > max_num:
                max_num = self.items[i]
        
        return max_num
    
    
    def remove(self):
        if self.__isEmpty():
            return -1
        self.items[self.count - 1] = 0
        self.count -= 1
    
    
    def removeAt(self, index):
        for i in range(index, self.count):
            self.items[i] = self.items[i+1]
        self.count -= 1
        
        
    def make_array(self, new_cap): 
        """ 
        Returns a new array with new_cap capacity 
        """ 
        return new_cap * [0] 
    
    
    def print_out(self):
        for i in range(self.count):
            print(self.items[i], end =" ")
        print(" ")


if __name__ == '__main__':
    arr = DynamicArray() 
    
    arr.insert(1) 
    arr.insert(2) 
    arr.insert(3) 
    arr.insert(4) 
    arr.insert(5) 
    arr.insertAt(6, 1) 
    
    arr.print_out()
    
    arr.reverse()
    arr.print_out()
    
    print(arr.max())
    
    arr.remove()
    arr.print_out()
    
    arr.removeAt(3)
    arr.print_out()
